pypit/arcoadd.py

```diff
@@ -460,66 +461,92 @@ def clean_cr(spectra, smask, n_grow_mask=1, nsig=5., nrej_low=5., debug=False, *
-        if n_grow_mask > 0:
-            cr1 = grow_mask(cr1, n_grow=n_grow_mask)
-        spectra.add_to_mask(cr1)
-        msgs.info("Rejecting {:d} CRs in exposure 1".format(np.sum(cr1)))
-        '''
-        # b-spline approach
-        # Package Data for convenience
-        waves = spectra.data['wave'].flatten()  # Packed 0,1
-        flux = fluxes.flatten()
-        sig = sigs.flatten()
-        #
-        gd = np.where(sig > 0.)[0]
-        srt = np.argsort(waves[gd])
-        idx = gd[srt]
-        mask, spl = arutils.robust_polyfit(waves[idx], flux[idx], 3, function='bspline',
-                weights=1./sig[gd][srt], sigma=3., maxone=False, everyn=6)
-        # Reject CR (with grow)
-        spec_fit = arutils.func_val(spl, wave, 'bspline')
-        for ii in range(2):
-            diff = fluxes[ii,:] - spec_fit
-            cr = (diff > nsig*sigs[ii,:]) & (sigs[ii,:]>0.)
-            if False:
-                debugger.plot1d(spectra.data['wave'][0,:], spectra.data['flux'][ii,:], spec_fit, xtwo=spectra.data['wave'][0,cr], ytwo=spectra.data['flux'][ii,cr], mtwo='s')
+        if cr_two_alg == 'diff': # b-spline approach
+            diff = fluxes[0,:] - fluxes[1,:]
+            # Robust mean/median
+            med, mad = arutils.robust_meanstd(diff)
+            # Spec0?
+            cr0 = (diff-med) > cr_nsig*mad
+            if n_grow_mask > 0:
+                cr0 = grow_mask(cr0, n_grow=n_grow_mask)
+            msgs.info("Rejecting {:d} CRs in exposure 0".format(np.sum(cr0)))
+            smask[0,cr0] = True
+            if debug:
+                debugger.plot1d(wave, fluxes[0,:], xtwo=wave[cr0], ytwo=fluxes[0,cr0], mtwo='s')
+            # Spec1?
+            cr1 = (-1*(diff-med)) > cr_nsig*mad
+            if n_grow_mask > 0:
+                cr1 = grow_mask(cr1, n_grow=n_grow_mask)
+            smask[1,cr1] = True
+            if debug:
+                debugger.plot1d(wave, fluxes[1,:], xtwo=wave[cr1], ytwo=fluxes[1,cr1], mtwo='s')
+            msgs.info("Rejecting {:d} CRs in exposure 1".format(np.sum(cr1)))
+        elif cr_two_alg == 'ratio': # b-spline approach
+            diff = fluxes[0,:] - fluxes[1,:]
+            rtio = fluxes[0,:] / fluxes[1,:]
+            # Robust mean/median
+            rmed, rmad = arutils.robust_meanstd(rtio)
+            dmed, dmad = arutils.robust_meanstd(diff)
+            # Spec0?            med, mad = arutils.robust_meanstd(diff)
+            cr0 = ((rtio-rmed) > cr_nsig*rmad) & ((diff-dmed) > cr_nsig*dmad)
+            if n_grow_mask > 0:
+                cr0 = grow_mask(cr0, n_grow=n_grow_mask)
+            msgs.info("Rejecting {:d} CRs in exposure 0".format(np.sum(cr0)))
+            smask[0,cr0] = True
+            if debug:
+                debugger.plot1d(wave, fluxes[0,:], xtwo=wave[cr0], ytwo=fluxes[0,cr0], mtwo='s')
+            # Spec1?
+            cr1 = (-1*(rtio-rmed) > cr_nsig*rmad) & (-1*(diff-dmed) > cr_nsig*dmad)
```

Comment:
The -1 are to identify CR's in the 2nd exposure (here indexed as 1)
A CR in that exposure will yield a large negative diff value.  Hence
the -1 is applied to make it positive.